src/jit/codegenxarch.cpp

```diff
@@ -1412,6 +1412,249 @@ void CodeGen::genCodeForBinary(GenTree* treeNode)
+{
+    assert(treeNode->OperGet() == GT_RETURN);
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+    return varTypeIsStruct(treeNode) ||
+           (treeNode->TypeGet() == TYP_VOID && compiler->info.compRetBuffArg != BAD_VAR_NUM);
+#else // !FEATURE_UNIX_AMD64_STRUCT_PASSING
+    assert(!varTypeIsStruct(treeNode));
+    return false;
+#endif // FEATURE_UNIX_AMD64_STRUCT_PASSING
+}
+
+//------------------------------------------------------------------------
+// genStructReturn: Generates code for returning a struct.
+//
+// Arguments:
+//    treeNode - The GT_RETURN tree node.
+//
+// Return Value:
+//    None
+//
+void
+CodeGen::genStructReturn(GenTreePtr treeNode)
+{
+    assert(treeNode->OperGet() == GT_RETURN);
+    GenTreePtr op1 = treeNode->gtGetOp1();
+    var_types targetType = treeNode->TypeGet();
+
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+    if (targetType == TYP_VOID)
+    {
+        assert(op1 == nullptr);
+        if (compiler->info.compRetBuffArg != BAD_VAR_NUM)
+        {
+            // System V AMD64 spec requires that when a struct is returned by a hidden
+            // argument the RAX should contain the value of the hidden retbuf arg.
+            getEmitter()->emitIns_R_S(INS_mov, EA_BYREF, REG_RAX, compiler->info.compRetBuffArg, 0);
+        }        
+    }
+    else
+    {
+        noway_assert((op1->OperGet() == GT_LCL_VAR) ||
+                     (op1->OperGet() == GT_CALL));
+
+        if (op1->OperGet() == GT_LCL_VAR)
+        {
+            assert(op1->isContained());
+
+            GenTreeLclVarCommon* lclVarPtr = op1->AsLclVarCommon();
+            LclVarDsc* varDsc = &(compiler->lvaTable[lclVarPtr->gtLclNum]);
+            assert(varDsc->lvDontPromote);
+
+            CORINFO_CLASS_HANDLE typeHnd = varDsc->lvVerTypeInfo.GetClassHandle();
+            assert(typeHnd != nullptr);
+
+            SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR structDesc;
+            compiler->eeGetSystemVAmd64PassStructInRegisterDescriptor(typeHnd, &structDesc);
+            assert(structDesc.passedInRegisters);
+            assert(structDesc.eightByteCount == CLR_SYSTEMV_MAX_EIGHTBYTES_COUNT_TO_PASS_IN_REGISTERS);
+
+            regNumber retReg0 = REG_NA;
+            unsigned __int8 offset0 = 0;
+            regNumber retReg1 = REG_NA;
+            unsigned __int8 offset1 = 0;
+
+            var_types type0 = TYP_UNKNOWN;
+            var_types type1 = TYP_UNKNOWN;
+
+            getStructTypeOffset(structDesc, &type0, &type1, &offset0, &offset1);
+            getStructReturnRegisters(type0, type1, &retReg0, &retReg1);
+
+            // Move the values into the return registers.
+            // 
+
+            assert(retReg0 != REG_NA && retReg1 != REG_NA);
+
+            getEmitter()->emitIns_R_S(ins_Load(type0), emitTypeSize(type0), retReg0, lclVarPtr->gtLclNum, offset0);
+            getEmitter()->emitIns_R_S(ins_Load(type1), emitTypeSize(type1), retReg1, lclVarPtr->gtLclNum, offset1);
+        }
+
```

Comment:
If op1 could be a call, then in else-part of this if stmnt we should consume op1 and move to RAX and RDX if not already in the right register(s)

else
{
   genConsumeRegs(op1);
   if (op1->gtRegNum != retReg0)
   {
        inst_RV_RV(INS_mov, retReg0, op1->gtRegNum, type0);
   }

// TODO - right now register specification of GT_RETURN/GT_CALL doesn't specify register pair for struct return
// in two registers.  Once that is done, this needs to be fixed to move 
// inst_RV_RV(INS_mov,  retReg1, op1->gtOtherRegNum, type1);
}

We need to keep the logic that obtains type0/type1/retReg0/retReg1 common to GT_LCL_VAR and GT_CALL case.